controllers/core/event_controller.go

```diff
@@ -0,0 +1,235 @@
+			Help: "The number of SGP events that were reconcile by the controller",
+		},
+		[]string{"operation"},
+	)
+
+	eventControllerSGPEventsFailureCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "failed_security_group_pod_events_reconciled_by_event_controller",
+			Help: "The number of SGP events that were reconcile by the controller but failed on label node",
+		},
+		[]string{"operation"},
+	)
+
+	eventControllerCNEventsCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "custom_networking_events_reconciled_by_event_controller",
+			Help: "The number of custom networking events that were reconcile by the controller",
+		},
+		[]string{"operation"},
+	)
+
+	eventControllerCNEventsFailureCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "failed_custom_networking_events_reconciled_by_event_controller",
+			Help: "The number of custom networking events that were reconcile by the controller but failed on label node",
+		},
+		[]string{"operation"},
+	)
+
+	eventControllerEventFailureCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "failed_get_events__by_event_controller",
+			Help: "The number of failures on getting events by the controller",
+		},
+		[]string{"operation"},
+	)
+
+	eventReconcileOperationLatency = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "event_reconcile_operation_latency",
+			Help: "Event controller reconcile events latency in ms",
+		},
+		[]string{"time"},
+	)
+
+	prometheusRegistered = false
+	reconcileEvents      = "reconcile_events"
+)
+
+func PrometheusRegister() {
+	if !prometheusRegistered {
+		metrics.Registry.MustRegister(eventControllerTotalEventsCount)
+		metrics.Registry.MustRegister(eventControllerSGPEventsCount)
+		metrics.Registry.MustRegister(eventControllerEventFailureCount)
+		metrics.Registry.MustRegister(eventControllerCNEventsCount)
+		metrics.Registry.MustRegister(eventReconcileOperationLatency)
+		metrics.Registry.MustRegister(eventControllerSGPEventsFailureCount)
+		metrics.Registry.MustRegister(eventControllerCNEventsFailureCount)
+		prometheusRegistered = true
+	}
+}
+
+func (r *EventReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	start := time.Now()
+	var err error
+
+	logger := r.Log.WithValues("event", req.Name)
+
+	eventControllerTotalEventsCount.WithLabelValues("reconcile").Inc()
+
+	ops := []client.ListOption{
+		client.MatchingFields{
+			EventFilterKey: config.VpcCNIReportingAgent,
+		},
+	}
+
+	events, err := r.K8sAPI.ListEvents(ops)
+	if err != nil {
+		logger.Error(err, "List events failed")
+		eventControllerEventFailureCount.WithLabelValues("failing_list").Inc()
+		return ctrl.Result{}, err
+	}
+
+	for _, event := range events.Items {
+		// only check and process events were created in the past EventCreatedPastMinutes minutes
+		if !event.GetCreationTimestamp().After(time.Now().Add(-time.Minute * EventCreatedPastMinutes)) {
```

Comment:
We seemed to be picking events in past 2 mins, Is this number based on scale testing? Also if event is continuously generated by the same node don't we need to skip those?

Reply:
Yes, the test was scaled to 1000 nodes. We should be safe for even larger clusters since VPC CNI will keep sending event till trunk is attached. Actually the node added by RC will no longer be processed since this code
```
func (k *k8sWrapper) AddLabelToManageNode(node *v1.Node, labelKey string, labelValue string) error {
	if node.Labels[labelKey] == labelValue {
		return nil
	} else {
		newNode := node.DeepCopy()
		newNode.Labels[labelKey] = labelValue
		return k.cacheClient.Status().Patch(k.context, newNode, client.MergeFrom(node))
	}
}
```